### PR TITLE
openshift_secondary_metadata_parser: Add 'version' parsing

### PR DESCRIPTION
**Pull request by wking**
Consuming the `version` file that graph-data grew in openshift/cincinnati-graph-data#233, so the logs will contain more useful error messages when the plugin is fed an incompatible schema version.

[1]: https://github.com/openshift/cincinnati-graph-data/pull/233

### CODE REVIEW

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (vrutkovs) [inline cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Lets declare this `pub` (same for `supported_version`) so that it could be reused in porting cincinnati-graph-data to Rust

Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Wouldn't we just run this whole plugin in the one-shot CI tool?

**Review by vrutkovs — CHANGES_REQUESTED** [review]
A better idea for `iterate and check` loop:

```
        let str_version = String::from_utf8_lossy(&version);
        if supported_versions.contains(&str_version.trim()) {
            Ok(())
        } else {
            Err(format_err!(
                "unrecognized graph-data version {}; supported versions: {:?}",
                str_version,
                supported_versions
            ))
        }
```

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (vrutkovs) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: yeah, but it needs to be public so that we could import it in `cincinnati-graph-data` CI too

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (vrutkovs) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: It would be best to have `pub static SUPPORTED_VERSIONS: Vec<&str> = vec!["1.0.0"];` at the head instead

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (steveej) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: @vrutkovs In the case which @wking describes we wouldn't need to use any code directly, but we'd run an appropriately configured graph-builder instead.

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (LalatenduMohanty) [inline cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: This array should not be local to this function. This was it looks like the version is hard coded.  The variable name should be metadata_schema_version.

**Review by LalatenduMohanty — CHANGES_REQUESTED** [review]

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (vrutkovs) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: so, eventually, in cincinnati-graph-data we'd run one-shot graph-builder? Sounds good to me, its a long shot though

Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Is a one-shot invocation of this plugin more achievable?

Comment (steveej) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: From the plugin perspective there's no other mode as one-shot as they don't compile to standalone binaries. It needs an embedding binary which in our case could be the graph-builder or the policy-engine. Just to mention it, I've toyed with the idea to make all of these the same binary with several sub-commands.

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Picked up the plural in cf1c8089 -> 7179eefe.  Will roll the public vector into [this thread][1].

[1]: https://github.com/openshift/cincinnati/pull/314#discussion_r499784365

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Does it need to be async?  There should be nothing outside of some local CPU/memory access in this code-path to cause any delays.

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: > This array should not be local to this function.

I would like to keep the set of supported versions internal, because I do not consider it a public API.  Folks who want to know if a graph-data checkout is supported should just load it with this plugin, and report back any error messages they see.  More on whether that's a viable option today vs. the current CI executable in [this thread][1], so I'm closing this to consolidate discussion there.

[1]: https://github.com/openshift/cincinnati/pull/314#discussion_r499784365

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Fixed with 7179eefe -> 1b34d2cf.

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Fixed the expected return type with c9ef3002 -> 2baa1f2f.

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Fixed in c9ef3002 -> 2baa1f2f.

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Fixed in c9ef3002 -> 2baa1f2f.

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: ah, the file read is async, so yes, it does.

Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Updated to use `async` and [`contains`][1] in 2baa1f2f -> 970b3183.

[1]: https://github.com/openshift/cincinnati/pull/314#pullrequestreview-503040937

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (vrutkovs) [inline cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Lets make it a top-level `const`?

**Review by vrutkovs — COMMENTED** [review]
looks good so far, one nit-pick

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: This is the same public API vs. internal method discussion we're having in [this existing thread][1], right?

[1]: https://github.com/openshift/cincinnati/pull/314#discussion_r499784365

Comment (vrutkovs) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Yes, however it would be easier to update this list ~~if~~ when we add new version there, even if its private

Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: Took a stab at this with 3b3e674c.

File: cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs
Comment (vrutkovs) [inline cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]:
```suggestion
static SUPPORTED_VERSIONS: &[&str] = &["1.0.0"];
```
rustfmt insists on a single line

Comment (wking) [reply cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs]: grr, rustfmt, making it slightly more tedious for folks to expand this list later :p  Fixed with 3b3e674c -> 89c15ecc.

**Review by steveej — COMMENTED** [review]
Some comments on the changes to the `graph_comparison_verbose`. Note that _verbose_ is in the name so the verbose output is by design.

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]:
```console
$ git --no-pager log -1
commit b7c4ac5a109f881d662ba6d611115bac458160c0 (HEAD -> graph-data-version, wking/graph-data-version)
Author: W. Trevor King <wking@tremily.us>
Date:   Fri Oct 30 10:15:46 2020 -0700

    cincinnati/src/lib: Ignore node order in graph comparison
    
    Node order has no semantic meaning, but edges do link nodes by index.
    To avoid false-positives like [1]:
    
      $ curl -s https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_cincinnati/314/pull-ci-openshift-cincinnati-master-e2e/1322037538348601344/build-log.txt | less
      ...
      ---- e2e_channel_success::_stable_4_2_s390x_ stdout ----
      Querying http://cincinnati-policy-engine-cincinnati-e2e.apps.ci-op-vdj2nvc4-90ac5.origin-ci-int-aws.dev.rhcloud.com/api/upgrades_info/v1/graph?channel=stable-4.2&arch=s390x
      thread 'e2e_channel_success::_stable_4_2_s390x_' panicked at 'Graphs differ! Showing differences from expected to actual', e2e/tests/e2e.rs:68:9
      stack backtrace:
      ...
    
    compare the nodes (after removing unwanted metadata keys) and edges
    (after converting to version-name indexing), instead of comparing the
    graphs directly.
    
    [1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/314/pull-ci-openshift-cincinnati-master-e2e/1322037538348601344
```

I dunno why `Graph:eq` wasn't happy with that, but it said the graphs differed and then listed no differences.

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: it was hard for me to understand whether the left or right was the expected or actual content.  And I'm still confused.  For example [this run][1] for 5ccd848 has:

```
thread 'e2e_channel_success::_stable_4_2_amd64_' panicked at 'Graphs differ! Showing differences from expected to actual
edges: 
-{
-  "4.1.20": [
-    "4.1.21",
-    "4.1.22",
-    "4.1.23",
-    "4.1.24",
-    "4.1.25",
-    "4.1.26",
-    "4.1.27",
-    "4.1.28",
-    "4.1.29"
-  ],
...
```

but the edges out of 4.1.20 were in the actual output but [_not_ part of the expected fixture][2] (so I'd expect them to have the `+` sense for "changes you'd want to apply to the test fixture to match the actual result").

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cincinnati/314/pull-ci-openshift-cincinnati-master-e2e/1322037538348601344
[2]: https://github.com/openshift/cincinnati/commit/5ccd84819f1e0eaebb62dfdf5a45ff49aa0ab78b#diff-5ba72bf7df7bf369e7a8e3263281c9349f4760045c0bcdccaa723c07cf7f79afR412

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: conversation continued in https://github.com/openshift/cincinnati/pull/314#discussion_r515282649

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: conversation consolidated into https://github.com/openshift/cincinnati/pull/314#discussion_r515284194

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: conversation consolidated into https://github.com/openshift/cincinnati/pull/314#discussion_r515284194

File: cincinnati/src/lib.rs
Comment (steveej) [reply cincinnati/src/lib.rs]: I would not completely rule out a bug in `Graph::eq` but on suspicion I'd prefer to not close our eyes and walk away from it ;-) Let's see if we can reproduce it. Would you mind extracting the text fixture bump into a separate PR or do you prefer me doing it?

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: there are a few changes made by `cincinnati/src/lib: Only print divergent portions of the diff` (currently 6c9accbe).  Instead of running separate threads on all the lines it touches, we can use this thread to debate whether we want to keep the commit or drop it.  I added it because the ANSI color coding out of `prettydiff::diff_lines` is not particularly convenient for me to consume, and between the effort of dumping the build log into a terminal to get the colors rendered and the false-positives around graph differences before 419246b03f (cincinnati/src/lib: Ignore node order in graph comparison, [thread][1]), it was easier for me to find the remaining bits of the fixtures I needed to touch if I did not print out the matching sections.  I am agnostic about whether we land this commit in the final PR, but it is helping me with the fixture iterations, so even if the consensus is "drop it before we merge the PR", I will keep it around until the fixtures are green.

[1]: https://github.com/openshift/cincinnati/pull/314#discussion_r515282649

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: Pushed as #343.

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: Fixed with 0e5110d2 in e861fa57

**Comment by wking** [comment]
Green :).  I've dropped the diff edits and moved to just the `version` handling and fixture bump with bd5d7e3f -> eaad2cbe.

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: I've [dropped][1] this change.

[1]: https://github.com/openshift/cincinnati/pull/314#issuecomment-720002421

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: I've [dropped][1] this change.

[1]: https://github.com/openshift/cincinnati/pull/314#issuecomment-720002421

File: cincinnati/src/lib.rs
Comment (wking) [reply cincinnati/src/lib.rs]: I've [dropped][1] this change.

[1]: https://github.com/openshift/cincinnati/pull/314#issuecomment-720002421

**Comment by wking** [comment]
All green, and I think I've addressed all the earlier review feedback or dropped the changes being reviewed ;).  One remaining thread is "[public or not?][1]", and I'd rather punt on that for now, and see if we can get the one-shot plugin approach going for graph-data CI.

/assign @steveeJ @vrutkovs 

[1]: https://github.com/openshift/cincinnati/pull/314/#discussion_r499784365

**Comment by steveej** [comment]
I'd find it cleaner if we split the version bump at least into a separate commit, and preferably accomplish it by merging the separate https://github.com/openshift/cincinnati/pull/343 first.

**Comment by wking** [comment]
Rebased on top of #343 with eaad2cbe -> 60a9aca3.

**Comment by steveej** [comment]
@wking do you want to merge this as is then or wait until and rebase again when #343 is merged?

**Review by steveej — APPROVED** [review]
/lgtm

**Pull request merged**
